filters/builtin/compress.go

```diff
@@ -0,0 +1,323 @@
+// the Content-Length value triggering the proxy to always return the response
+// with chunked transfer encoding, sets the Content-Encoding to the selected
+// encoding and sets the Vary: Accept-Encoding header, if missing.
+//
+// The compression happens in a streaming way, using only a small internal buffer.
+//
+func NewCompress() filters.Spec { return &compress{} }
+
+func (c *compress) Name() string {
+	return CompressName
+}
+
+func (c *compress) CreateFilter(args []interface{}) (filters.Filter, error) {
+	f := &compress{}
+
+	if len(args) == 0 {
+		f.mime = defaultCompressMIME
+		return f, nil
+	}
+
+	if args[0] == "..." {
+		f.mime = defaultCompressMIME
+		args = args[1:]
+	}
+
+	for _, a := range args {
+		if s, ok := a.(string); ok {
+			f.mime = append(f.mime, s)
+		} else {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	return f, nil
+}
+
+func (c *compress) Request(_ filters.FilterContext) {}
+
+func stringsContain(ss []string, s string, transform ...func(string) string) bool {
+	for _, si := range ss {
+		for _, t := range transform {
+			si = t(si)
+		}
+
+		if si == s {
+			return true
+		}
+	}
+
+	return false
+}
+
+func canEncodeEntity(r *http.Response, mime []string) bool {
+	if ce := r.Header.Get("Content-Encoding"); ce != "" && ce != "identity" /* forgiving for identity */ {
+		return false
+	}
+
+	cc := strings.Split(r.Header.Get("Cache-Control"), ",")
+	if stringsContain(cc, "no-transform", strings.TrimSpace, strings.ToLower) {
+		return false
+	}
+
+	ct := r.Header.Get("Content-Type")
+	if i := strings.Index(ct, ";"); i >= 0 {
+		ct = ct[:i]
+	}
+
+	if !stringsContain(mime, ct) {
+		return false
+	}
+
+	return true
+}
+
+func acceptedEncoding(r *http.Request) string {
+	var encs encodings
+	for _, s := range strings.Split(r.Header.Get("Accept-Encoding"), ",") {
+		sp := strings.Split(s, ";")
+		if len(sp) == 0 {
+			continue
+		}
+
+		name := strings.ToLower(strings.TrimSpace(sp[0]))
+		if !stringsContain(supportedEncodings, name) {
+			continue
+		}
+
+		enc := &encoding{name, 1}
+		encs = append(encs, enc)
+
+		for _, spi := range sp[1:] {
+			spi = strings.TrimSpace(spi)
+			if !strings.HasPrefix(spi, "q=") {
+				continue
+			}
+
+			q, err := strconv.ParseFloat(strings.TrimPrefix(spi, "q="), 32)
+			if err != nil {
+				continue
+			}
+
+			enc.q = float32(q)
+			break
+		}
+	}
+
+	if len(encs) == 0 {
+		return ""
+	}
+
+	sort.Sort(encs)
+	return encs[0].name
+}
+
+func responseHeader(r *http.Response, enc string) {
+	r.Header.Del("Content-Length")
+	r.Header.Set("Content-Encoding", enc)
+
+	if !stringsContain(r.Header["Vary"], "Accept-Encoding", http.CanonicalHeaderKey) {
+		r.Header.Add("Vary", "Accept-Encoding")
+	}
+}
+
+// Not handled encoding is considered as an implementation error, since
+// these functions are only called from inside the package, and the
+// encoding should be selected from a predefined set.
+func unsupported() {
+	panic(unsupportedEncoding)
```

Comment:
I understand the reasoning, but I'm not sure I like the idea of a filter panicking...